hibernate-core/src/main/java/org/hibernate/type/descriptor/java/LocaleTypeDescriptor.java

```diff
@@ -51,43 +51,33 @@ public Locale fromString(String string) {
 			return Locale.ROOT;
 		}
 
-		boolean separatorFound = false;
-		int position = 0;
-		char[] chars = string.toCharArray();
-
-		for ( int i = 0; i < chars.length; i++ ) {
-			// We just look for separators
-			if ( chars[i] == '_' ) {
-				if ( !separatorFound ) {
-					// On the first separator we know that we have at least a language
-					string = new String( chars, position, i - position );
-					position = i + 1;
-				}
-				else {
-					// On the second separator we have to check whether there are more chars available for variant
-					if ( chars.length > i + 1 ) {
-						// There is a variant so add it to the constructor
-						return new Locale( string, new String( chars, position, i - position ), new String( chars,
-								i + 1, chars.length - i - 1 ) );
+		Locale.Builder builder = new Locale.Builder();
+		String[] parts = string.split("_");
+
+		for (int i = 0; i < parts.length; i++) {
+			String s = parts[i];
+			switch (i) {
+				case 0:
+					builder.setLanguage(s);
+					break;
+				case 1:
+					builder.setRegion(s);
+					break;
+				case 2:
+					if (i < parts.length - 1 || !s.startsWith("#")) {
+						builder.setVariant(s);
+						break;
 					}
-					else {
-						// No variant given, we just have language and country
-						return new Locale( string, new String( chars, position, i - position ), "" );
+				case 3:
+					if (s.startsWith("#")) {
```

Comment:
When needing to compare with characters, I'd prefer to use the char API:
    string.charAt(0) == '#'